File: src/report.ts
import type { BenchmarkManifest, BenchmarkResults, OperationName, OperationRun } from './types.js';

function formatMs(value: number): string {
  return value.toFixed(2);
}

function formatRatio(value: number): string {
  if (!Number.isFinite(value) || value === 0) {
    return '-';
  }
  return `${value.toFixed(2)}x`;
}

function groupByOperation(results: BenchmarkResults['results']): Map<OperationName, BenchmarkResults['results']> {
  const grouped = new Map<OperationName, BenchmarkResults['results']>();

  for (const result of results) {
    const current = grouped.get(result.operation) ?? [];
    current.push(result);
    grouped.set(result.operation, current);
  }

  return grouped;
}

export function renderSummary(
  manifest: BenchmarkManifest,
  benchmarkResults: BenchmarkResults,
  operationRuns: OperationRun[] = [],
): string {
  const lines: string[] = [];

  lines.push(`# Tigris benchmark summary`);
  lines.push('');
  lines.push(`Run ID: \`${manifest.runId}\``);
  lines.push(`Created: ${benchmarkResults.createdAt}`);
  lines.push(`Region placement: \`single\` / \`iad\``);
  lines.push(`Object size: ${manifest.options.objectSizeBytes} bytes`);
  lines.push(`Iterations: ${manifest.options.iterations}`);
  lines.push(`Warmup: ${manifest.options.warmup}`);
  lines.push(`Concurrency: ${manifest.options.concurrency}`);
  lines.push('');
  lines.push('Isolation: each operation is provisioned and benchmarked against a fresh bucket set. Bucket creation, snapshot creation, and fork creation are not included in measured latency.');
  lines.push('');
  lines.push('Scenario map:');
  for (const scenario of manifest.scenarios) {
    const parent = scenario.parentScenarioId ? `, parent=${scenario.parentScenarioId}` : '';
    lines.push(`- ${scenario.id}: snapshotEnabled=${scenario.snapshotEnabled}, forkDepth=${scenario.forkDepth}${parent}`);
  }
  lines.push('');

  if (operationRuns.length > 0) {
    lines.push('Per-operation runs:');
    for (const operationRun of operationRuns) {
      lines.push(`- ${operationRun.operation}: run=\`${operationRun.runId}\`, artifacts=\`${operationRun.artifactDir}\``);
    }
    lines.push('');
  }

  const grouped = groupByOperation(benchmarkResults.results);

  for (const [operation, results] of grouped.entries()) {
    const sorted = [...results].sort((left, right) => {
      const scenarioLeft = manifest.scenarios.find((scenario) => scenario.id === left.scenarioId);
      const scenarioRight = manifest.scenarios.find((scenario) => scenario.id === right.scenarioId);
      return (scenarioLeft?.forkDepth ?? 0) - (scenarioRight?.forkDepth ?? 0) || left.scenarioId.localeCompare(right.scenarioId);
    });
    const normal = sorted.find((result) => result.scenarioId === 'normal');

    lines.push(`## ${operation}`);
    lines.push('');
    lines.push('| Scenario | p50 ms | p95 ms | p99 ms | mean ms | ops/s | p50 vs normal |');
    lines.push('| --- | ---: | ---: | ---: | ---: | ---: | ---: |');

    for (const result of sorted) {
      const ratio = normal ? result.stats.p50Ms / normal.stats.p50Ms : 0;
      lines.push(
        `| ${result.scenarioId} | ${formatMs(result.stats.p50Ms)} | ${formatMs(result.stats.p95Ms)} | ${formatMs(
          result.stats.p99Ms,
        )} | ${formatMs(result.stats.meanMs)} | ${result.stats.opsPerSecond.toFixed(2)} | ${formatRatio(ratio)} |`,
      );
    }

    lines.push('');
  }

  return `${lines.join('\n')}\n`;
}

File: gobench/report.go
package gobench

import (
	"fmt"
	"sort"
	"strings"
)

func RenderSummary(manifest BenchmarkManifest, benchmarkResults BenchmarkResults, operationRuns []OperationRun, implementation string) string {
	lines := []string{
		"# Tigris benchmark summary",
		"",
		fmt.Sprintf("Run ID: `%s`", manifest.RunID),
		fmt.Sprintf("Created: %s", benchmarkResults.CreatedAt),
		"Region placement: `single` / `iad`",
		fmt.Sprintf("Object size: %d bytes", manifest.Options.ObjectSizeBytes),
		fmt.Sprintf("Iterations: %d", manifest.Options.Iterations),
		fmt.Sprintf("Warmup: %d", manifest.Options.Warmup),
		fmt.Sprintf("Concurrency: %d", manifest.Options.Concurrency),
		fmt.Sprintf("Implementation: `%s`", implementation),
		"",
		"Isolation: each operation is provisioned and benchmarked against a fresh bucket set. Bucket creation, snapshot creation, and fork creation are not included in measured latency.",
		"",
		"Scenario map:",
	}

	for _, scenario := range manifest.Scenarios {
		parent := ""
		if scenario.ParentScenarioID != "" {
			parent = fmt.Sprintf(", parent=%s", scenario.ParentScenarioID)
		}
		lines = append(lines, fmt.Sprintf("- %s: snapshotEnabled=%t, forkDepth=%d%s", scenario.ID, scenario.SnapshotEnabled, scenario.ForkDepth, parent))
	}
	lines = append(lines, "")

	if len(operationRuns) > 0 {
		lines = append(lines, "Per-operation runs:")
		for _, run := range operationRuns {
			lines = append(lines, fmt.Sprintf("- %s: run=`%s`, artifacts=`%s`", run.Operation, run.RunID, run.ArtifactDir))
		}
		lines = append(lines, "")
	}

	grouped := map[OperationName][]OperationResult{}
	for _, result := range benchmarkResults.Results {
		grouped[result.Operation] = append(grouped[result.Operation], result)
	}

	scenarioDepth := map[string]int{}
	for _, scenario := range manifest.Scenarios {
		scenarioDepth[scenario.ID] = scenario.ForkDepth
	}

	for _, operation := range OperationNames {
		results := grouped[operation]
		if len(results) == 0 {
			continue
		}
		sort.Slice(results, func(i, j int) bool {
			leftDepth := scenarioDepth[results[i].ScenarioID]
			rightDepth := scenarioDepth[results[j].ScenarioID]
			if leftDepth != rightDepth {
				return leftDepth < rightDepth
			}
			return results[i].ScenarioID < results[j].ScenarioID
		})

		var normal *OperationResult
		for idx := range results {
			if results[idx].ScenarioID == "normal" {
				normal = &results[idx]
				break
			}
		}

		lines = append(lines,
			fmt.Sprintf("## %s", operation),
			"",
			"| Scenario | p50 ms | p95 ms | p99 ms | mean ms | ops/s | p50 vs normal |",
			"| --- | ---: | ---: | ---: | ---: | ---: | ---: |",
		)

		for _, result := range results {
			ratio := "-"
			if normal != nil && normal.Stats.P50MS > 0 {
				ratio = fmt.Sprintf("%.2fx", result.Stats.P50MS/normal.Stats.P50MS)
			}
			lines = append(lines, fmt.Sprintf(
				"| %s | %.2f | %.2f | %.2f | %.2f | %.2f | %s |",
				result.ScenarioID,
				result.Stats.P50MS,
				result.Stats.P95MS,
				result.Stats.P99MS,
				result.Stats.MeanMS,
				result.Stats.OpsPerSecond,
				ratio,
			))
		}
		lines = append(lines, "")
	}

	return strings.Join(lines, "\n") + "\n"
}
